### PR TITLE
[Train] Add Intel HPU Support and Training Examples to Ray Train

### PR DESCRIPTION
**Pull request by kira-lin**
## Why are these changes needed?

To leverage the potential of Intel's Habana Processing Units (HPUs), we extend Ray Train's capabilities by adding support for Intel HPU hardware. This update also includes training scripts for BERT and ResNet models, which are widely used in NLP and computer vision tasks, respectively.

## Related issue number
Replace #42866

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by woshiyyya** [comment]
@kira-lin Can you address my comments in your previous PR https://github.com/ray-project/ray/pull/42866?

**Comment by kira-lin** [comment]
@woshiyyya I've addressed all of them, please review.

File: doc/source/train/examples/hpu/bert.ipynb [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+    "The `train_bert` function sets up the distributed training environment using Ray and starts the training process."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_bert(num_workers=2, use_gpu=False):\n",
```
Comment (woshiyyya) [inline doc/source/train/examples/hpu/bert.ipynb:205]: We are not using the `use_gpu` flag here? We can focus on HPU usage and just remove this option.

File: doc/source/train/examples/hpu/bert.ipynb [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+    "    print(f\"Training result: {result}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Start Training\n",
+    "\n",
+    "Finally, we call the `train_bert` function to start the training process. You can adjust the number of workers and whether to use a GPU."
```
Comment (woshiyyya) [inline doc/source/train/examples/hpu/bert.ipynb:239]: Remove the GPU mention here as well

**Review by woshiyyya — CHANGES_REQUESTED** [review]
@kira-lin Thanks for the hard work. The PR looks much better now. 

I left some comments on the example and please take a look. It should be good to go after we polished the wordings in the examples.

File: doc/source/train/examples/hpu/bert.ipynb [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\u001b[36m(pid=59382)\u001b[0m /usr/local/lib/python3.10/dist-packages/torch/distributed/distributed_c10d.py:252: UserWarning: Device capability of hccl unspecified, assuming `cpu` and `cuda`. Please specify it via the `devices` argument of `register_backend`.\n",
```
[the remark this answers is not in the document]
Comment (kira-lin) [reply doc/source/train/examples/hpu/bert.ipynb:337]: I mentioned this in the comment. The warning is due to habana driver 1.13.0 I was using. It gets fixed in the latest 1.14.0

**Comment by kira-lin** [comment]
@woshiyyya addressed

**Review by woshiyyya — APPROVED** [review]
LGTM. Thanks for the contribution! 

@kira-lin We will put the examples into Community Example section. In the future, if there is a user issue related to these examples, I'll hand it over to you.

File: python/ray/train/torch/config.py [hunk 4]
```diff
@@ -8,11 +8,16 @@
...
 import ray
+from ray._private.accelerators.hpu import HPU_PACKAGE_AVAILABLE
 from ray.train._internal.utils import get_address_and_port
 from ray.train._internal.worker_group import WorkerGroup
 from ray.train.backend import Backend, BackendConfig
 from ray.util import PublicAPI
 
+if HPU_PACKAGE_AVAILABLE:
+    import habana_frameworks.torch.core as htcore  # noqa: F401
+    import habana_frameworks.torch.distributed.hccl as hpu_dist  # noqa: F401
```
Comment (matthewdeng) [inline python/ray/train/torch/config.py:19]: Should this import logic be moved to within `_setup_torch_process_group` and only called when `backend == "hccl"`? Seems like this is needed to be called on the workers.

Comment (kira-lin) [reply python/ray/train/torch/config.py:19]: updated

Comment (matthewdeng) [reply python/ray/train/torch/config.py:19]: @kira-lin can you push the change to the PR?

Comment (kira-lin) [reply python/ray/train/torch/config.py]: sorry, forgot 🥲 pushed

File: python/ray/train/torch/config.py [hunk 5]
```diff
@@ -104,6 +105,9 @@ def _setup_torch_process_group(
             "To override this behavior, you can set NCCL_ASYNC_ERROR_HANDLING=0."
         )
         os.environ["NCCL_ASYNC_ERROR_HANDLING"] = "1"
+    elif HPU_PACKAGE_AVAILABLE and backend == "hccl":
```
Comment (matthewdeng) [inline python/ray/train/torch/config.py:108]: nit for readability: 

```suggestion
    elif backend == "hccl" and HPU_PACKAGE_AVAILABLE:
```

**Review by matthewdeng — APPROVED** [review]

File: doc/source/train/examples.rst
Comment (matthewdeng) [inline doc/source/train/examples.rst:1]: Looks like there is a merge conflict for this file, could you take a look at resolving it?

Comment (kira-lin) [reply doc/source/train/examples.rst:1]: resolved

File: doc/source/train/examples.yml [hunk 6]
```diff
@@ -47,6 +47,20 @@ examples:
       - horovod
     skill_level: beginner
     link: examples/horovod/horovod_example
+  - title: Train ResNet model with HPU
```
Comment (woshiyyya) [inline doc/source/train/examples.yml:50]:
```suggestion
  - title: [Community Example] Train ResNet model with HPU
```

File: doc/source/train/examples.yml [hunk 7]
```diff
@@ -47,6 +47,20 @@ examples:
...
+    frameworks:
+      - pytorch
+    skill_level: beginner
+    use_cases:
+      - computer vision
+    link: examples/hpu/resnet.ipynb
```
Comment (woshiyyya) [inline doc/source/train/examples.yml:56]:
```suggestion
    link: examples/hpu/resnet
```

File: doc/source/train/examples.yml [hunk 8]
```diff
@@ -47,6 +47,20 @@ examples:
...
+  - title: Train BERT model with HPU
+    frameworks:
+      - transformers
+    skill_level: beginner
+    use_cases:
+      - natural language processing
+    link: examples/hpu/bert.ipynb
```
Comment (woshiyyya) [inline doc/source/train/examples.yml:63]:
```suggestion
    link: examples/hpu/bert
```

File: doc/source/train/examples.yml [hunk 9]
```diff
@@ -47,6 +47,20 @@ examples:
...
     skill_level: beginner
     link: examples/horovod/horovod_example
+  - title: Train ResNet model with HPU
+    frameworks:
+      - pytorch
+    skill_level: beginner
+    use_cases:
+      - computer vision
+    link: examples/hpu/resnet.ipynb
+  - title: Train BERT model with HPU
```
Comment (woshiyyya) [inline doc/source/train/examples.yml:57]:
```suggestion
  - title: [Community Example] Train BERT model with HPU
```

**Review by woshiyyya — CHANGES_REQUESTED** [review]

File: doc/source/train/examples.yml
Comment (matthewdeng) [inline doc/source/train/examples.yml:1]: Looks like the doc build is failing. Would you be able to take a look? I can also take a look later if needed.

Comment (kira-lin) [reply doc/source/train/examples.yml:1]: sorry, not sure how to fix it

**Review by woshiyyya — APPROVED** [review]

**Comment by liuxsh9** [comment]
It's great to see that Ray Train has successfully extended its support beyond GPUs to other acceleration hardware. Interestingly, we are also preparing to contribute support for Huawei NPU. The adaptation process for torch_npu is very similar to torch_hpu. 
However, we are concerned that following the HPU approach to add NPU support might make the code more complex and introduce scattered device type and library checks. 
Therefore, in terms of design, we would like to separate the hardware-specific components, similar to the abstraction of `accelerators`, to make it easier for Ray Train (even RLlib) to support third-party devices. 
We have already made some designs that ensure it won't affect the usage of GPUs and HPUs, and we provide references for integrating third-party devices with APIs like `prepare_model` and `prepare_data_loader`. We would love to hear the community's opinions on this matter.
Should we initiate a PR to showcase the implementation (including some code migration for HPU), or would it be better to initiate an issue first to discuss this matter? @kira-lin @woshiyyya @matthewdeng

**Comment by woshiyyya** [comment]
Hi @liuxsh9 , you are welcome to add support for NPU accelerator in Ray Train: ) If you already have the designs, feel free to post a PR and we can discuss the implementation details.

**Comment by anyscalesam** [comment]
@angelinalg can we pick this into latest; I think it might have just missed the branch cut date so right now the "latest" URL renders to a 404.

Switching to master works fine though.

https://docs.ray.io/en/latest/train/examples/hpu/resnet.html
https://docs.ray.io/en/master/train/examples/hpu/resnet.html

**Pull request merged**